src/Codegen/DirectivesFinder.hack

```diff
@@ -0,0 +1,118 @@
+        // This requires building the qualified name, which is not
+        // available in `getAttributes`, unfortunately.
+        $attributes = $method->getAttributes();
+        foreach ($attributes as $attribute => $args) {
+            // Consider names starting with `\` to be fully qualified
+            if (!Str\starts_with($attribute, '\\')) {
+                //
+                // Try to determine the fully qualified name of the attribute.
+                //
+                // There are two cases here: 
+                // If the attribute is declared as `Baz`, then either:
+                //      1) Baz is not namespaced at all.
+                //      2) Baz is a imported via a use type statement, e.g., `use type Foo\Bar\Baz`.
+                // If the attribute is declared as `Bar/Baz`, then either:
+                //      1) Bar is a top-level namespace.
+                //      2) Bar is imported via a use namespace statement, e.g., `use namespace Foo\Bar`.
+                //
+                // In each case, we need to match the whatever comes before the first `\` in the attribute
+                // name (or the whole attribute name, if it contains no slashes) with whatever comes after
+                // the last `\` in each use type or use namespace statement.
+                //
+                // Once we have a match, we can build the fully qualified name by adding the attribute name
+                // to everything that comes before the last `\` in the eligible use statement.
+                //
+                $needle = Str\split($attribute, '\\')
+                    |> C\firstx($$);
+                $attribute = '\\'.$attribute;
+                foreach ($use_decls as $decl) {
+                    if (C\contains_key($decl['leaves'], $needle)) {
+                        // We have a match such as `Bar\Baz` with `use namespace Foo\Bar`;
+                        // build the qualified name as `Foo\Bar\Baz`.
+                        $attribute = '\\'.$decl['root'].$attribute;
+                        break;
+                    }
+                }
+            }
+
+            if (C\contains_key($this->directives, $attribute)) {
+                $directives[] = Str\format(
+                    'new %s(%s)',
+                    $attribute,
+                    Vec\map(
+                        $args,
+                        $arg ==> \var_export($arg, true)
+                            |> Str\replace($$, 'varray', 'vec')
+                            |> Str\replace($$, 'darray', 'dict')
+                            |> Str\replace($$, 'array', 'shape')
+                            |> Str\replace($$, 'vec ', 'vec')
+                            |> Str\replace($$, 'dict ', 'dict')
+                            |> Str\replace($$, 'shape ', 'shape')
+                    )
+                        |> Str\join($$, ', '),
+                );
+            }
+        }
+
+        return $directives;
```

Comment:
I recommend always generating some of the code as part of the PR so that people can see and review the generated code as if it were human-written. Generally speaking it's best to hold generated code to the same high standards as human-written code.

Reply:
No, I meant that I could not identify by looking at your link that it was a link to lower in this PR. :)

Reply:
Gotcha. I'd recommend reviewing this whole file: https://github.com/mwildehahn/hack-graphql/pull/123/files#diff-bc6416c0810e5e09cdcf817fbe7fc23de76a1884523575597eb7a6b4bc6b9861

It contains generated code for object and field-level directives.